azure-iot-hub/tests/test_iothub_registry_manager.py

```diff
@@ -0,0 +1,881 @@
+    @pytest.mark.parametrize(
+        "primary_key, secondary_key", testdata, ids=["Primary Key", "Secondary Key"]
+    )
+    def test_initializes_device_with_kwargs_for_sas(
+        self,
+        mock_service_operations,
+        iothub_registry_manager,
+        mock_module_constructor,
+        primary_key,
+        secondary_key,
+    ):
+        iothub_registry_manager.create_module_with_sas(
+            device_id=fake_device_id,
+            module_id=fake_module_id,
+            managed_by=fake_managed_by,
+            status=fake_status,
+            primary_key=primary_key,
+            secondary_key=secondary_key,
+        )
+
+        assert mock_module_constructor.call_count == 1
+
+        assert mock_module_constructor.call_args[1]["module_id"] == fake_module_id
+        assert mock_module_constructor.call_args[1]["managed_by"] == fake_managed_by
+        assert mock_module_constructor.call_args[1]["device_id"] == fake_device_id
+        assert mock_module_constructor.call_args[1]["status"] == fake_status
+        assert isinstance(
+            mock_module_constructor.call_args[1]["authentication"], AuthenticationMechanism
+        )
+        auth_mechanism = mock_module_constructor.call_args[1]["authentication"]
+        assert auth_mechanism.type == "sas"
+        assert auth_mechanism.x509_thumbprint is None
+        sym_key = auth_mechanism.symmetric_key
+        assert sym_key.primary_key == primary_key
+        assert sym_key.secondary_key == secondary_key
+
+    @pytest.mark.it(
+        "calls method from service operations with device id, module id and previously constructed module"
+    )
+    @pytest.mark.parametrize(
+        "primary_key, secondary_key", testdata, ids=["Primary Key", "Secondary Key"]
+    )
+    def test_calls_create_or_update_device_for_sas(
+        self,
+        mock_module_constructor,
+        mock_service_operations,
+        iothub_registry_manager,
+        primary_key,
+        secondary_key,
+    ):
+        iothub_registry_manager.create_module_with_sas(
+            device_id=fake_device_id,
+            status=fake_status,
+            module_id=fake_module_id,
+            managed_by=fake_managed_by,
+            primary_key=primary_key,
+            secondary_key=secondary_key,
+        )
+
+        assert mock_service_operations.create_or_update_module.call_count == 1
+        assert mock_service_operations.create_or_update_module.call_args[0][0] == fake_device_id
+        assert mock_service_operations.create_or_update_module.call_args[0][1] == fake_module_id
+        assert (
+            mock_service_operations.create_or_update_module.call_args[0][2]
+            == mock_module_constructor.return_value
+        )
+
+
+@pytest.mark.describe("IoTHubRegistryManager - .create_module_with_x509()")
+class TestCreateModuleWithX509(object):
+
+    testdata = [(fake_primary_thumbprint, None), (None, fake_secondary_thumbprint)]
```

Comment:
>testdata = [(fake_primary_thumbprint, None), (None, fake_secondary_thumbprint)] [](start = 4, length = 79)

If you wanted to preserve the ids as well instead of needing to duplicate them on every test you could add them here with the use of pytest.param

Reply:
I wanted to parametrize the test by doing on top of the test itself but I used test data so that it can be shared between the tests in a class...just tried out something different also...

---
In reply to: [332243127](https://github.com/Azure/azure-iot-sdk-python/pull/340#discussion_r332243127) [](ancestors = 332243127)